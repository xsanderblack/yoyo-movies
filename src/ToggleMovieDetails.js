import React, { Component } from 'react';
import axiosWithCache from './axiosWithCache';
import MovieDetailsScreen from './MovieDetailsScreen';
import {
  PATH_BASE,
  PATH_MOVIE,
  API_KEY
} from './Constants';

// Component renders button to show or hide movie details screen,
// makes request and passes the result to MovieDetailsScreen component.
export default class ToggleMovieDetails extends Component {
  state = {
    movie: null,
    isLoading: false,
    error: null
  };
  
  // Methods checks, whenever movie details is displayed and closes it, if yes.
  // If no, method makes request and populates the state with result.
  toggleMovieDetails = () => {
    const { movie, error } = this.state;

    if (error) this.setState({ error: null });
    this.setState({ isLoading: true });

    if (movie) {
      this.setState({ movie: null, isLoading: false });
    } else {
      axiosWithCache({
        url: `${PATH_BASE}${PATH_MOVIE}${this.props.id}?api_key=${API_KEY}`,
        method: 'get'
      })
        .then(result => {
          this.setState({ movie: result.data, isLoading: false });
        })
        .catch(error => this.setState({ error, isLoading: false }));
    };
  };

  render() {
    const { movie, isLoading, error } = this.state;

    return (
      <div>
        <button onClick={this.toggleMovieDetails}>
          {movie ? 'Hide Details' : 'Show Details'}
        </button>
        {movie && <MovieDetailsScreen {...movie} />}
        <div>
          {isLoading && 'Loading...'}
          {error && 'An error occured. Please, try again later.'}
        </div>
      </div>
    );
  };
};
