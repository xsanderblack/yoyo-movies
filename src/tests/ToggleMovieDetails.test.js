import React from 'react';
import ReactDOM from 'react-dom';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import ToggleMovieDetails from '../ToggleMovieDetails';

it('ToggleMovieDetails renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<ToggleMovieDetails />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test('should render ToggleMovieDetails correctly', () => {
  const renderer = new ReactShallowRenderer();
  renderer.render(<ToggleMovieDetails />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
