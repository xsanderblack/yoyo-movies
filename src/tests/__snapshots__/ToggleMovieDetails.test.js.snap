// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ToggleMovieDetails correctly 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Show Details
  </button>
  <div />
</div>
`;
